### PR TITLE
Add config options to control the editor mode

### LINKED ISSUE #642: Focus jumps to latest position when 2 tabs are open side by side
Same issue and behaviour as #427 but for editors opened in side-by-side tabs

### PR DESCRIPTION
**Pull request by QAston**
defaultMode - What mode to start the editor in.
resetModeOnWindowChange - "If 'true'
    will reset windows to the defaultMode on window change

Fixes #642 #613

### CODE REVIEW

**Comment by trkoch** [comment]
Any ideas why the test is failing? @QAston Is this ready for review?

File: package.json [hunk 1]
```diff
@@ -135,6 +135,20 @@
...
+                    "enum": [
+                        "insert",
+                        "normal"
+                    ],
+                    "description": "What mode to start the editor in."
+                },
+                "vscode-neovim.resetModeOnWindowChange": {
+                    "type": "boolean",
+                    "default": true,
+                    "markdownDescription": "If 'true' will reset windows to the vscode-neovim.defaultMode on window change."
```
Comment (trkoch) [inline package.json:150]: Just for clarification, window change means when you leave Vscode (e.g. switch to browser)?

File: package.json [hunk 2]
```diff
@@ -1260,6 +1274,7 @@
     },
     "scripts": {
         "lint": "eslint --ext .ts src",
+        "lint-fix": "eslint --fix --ext .ts src",
```
Comment (trkoch) [inline package.json:1277]: Personally I'd prefer unrelated changes like this to be in a separate PR.

File: package.json (hunk 1 above)
Comment (QAston) [reply package.json:150]: It means switching to a different editor in vscode (perhaps editor change is a better name)

**Comment by QAston** [comment]
@trkoch the test was failing on master - now the master is fixed I'm going to look into cleaning up and updating this PR (i've been happily using the changes for a while)

File: package.json (hunk 1 above)
Comment (trkoch) [reply package.json:150]: Seems like Vscode terminology is in fact _editor_ (as indicated eg. by the command _Move Editor into Right Group_).

Comment (QAston) [reply package.json:150]: Yeah, I might have been thinking about nvim's window, but I agree editor is better.

**Comment by QAston** [comment]
I've done some cleanups, but the ci still fails because it also fails on the master branch

**Comment by trkoch** [comment]
@theol0403 Can you give us a hint?

**Comment by justinmk** [comment]
> Fixes #642 #613

Adding an option doesn't fix anything. Either the issue is fixed or it isn't. Making users fiddle with options is a pretend fix. Especially with the amount of code added here.

Is it possible to do something reasonable for all users?

**Comment by QAston** [comment]
People have different preferences, some people like to be in normal mode all the time. This preserves the current default.

**Comment by justinmk** [comment]
So you can define a `VimEnter` autocmd that sets whatever mode you want. 

Why does there need to be a special option for this? The PR does not have a clear description. I glanced at the reference issues, they are too chatty. What is a 2-sentence explanation of the problem this PR solves?

**Comment by trkoch** [comment]
> VimEnter autocmd

In general I'm of course in favor to leverage default vim features where possible. However, I tried this and if I remember it correctly, it introduces a lag.

> What is a 2-sentence explanation of the problem this PR solves?

* `defaultMode` allows to start in insert mode, without the lag introduced by `VimEnter` autocmd
* `resetModeOnActiveEditorChange` controls whether mode is restored to normal when switching editors ("buffers" in vim terms). If set to `false` and currently in insert mode, the editor will stay in insert mode after you change the editor.

`resetModeOnActiveEditorChange` can be useful if you plan to stay in insert mode for an extended period. Why you would do this of course depends on your usage and preferences. Personally I found myself wanting this when clicking/selecting/pasting across editors.

**Comment by trkoch** [comment]
> Making users fiddle with options is a pretend fix.

I don't think it was meant as a clear-and-cut fix, rather a feature "fixing" a feature request (as it is often rather confusingly written).

> Especially with the amount of code added here.

@QAston Perhaps splitting both options into multiple PRs makes things easier to follow.

**Comment by QAston** [comment]
@justinmk I've tried doing autocmds, using them caused other issues

@trkoch I think I'm going to work on my own plugin instead 😛

**Comment by justinmk** [comment]
> @justinmk I've tried doing autocmds, using them caused other issues

Those are the issues we should focus on. Adding layers of special-case options bloats documentation, codebase, and avoids fixing actual problems. Entire purpose of having embedded Nvim is to use Nvim features.



> `resetModeOnActiveEditorChange` controls whether mode is restored to normal when switching editors ("buffers" in vim terms). If set to `false` and currently in insert mode, the editor will stay in insert mode after you change the editor.

That can be done with a `WinEnter` handler. I've just confirmed that this works:

    au WinEnter * startinsert

So `resetModeOnActiveEditorChange` absolutely is an anti-feature.

**Comment by QAston** [comment]
@justinmk Yes, it works but makes the cursor break with multiple editors if i recall correctly. Feel free to take over the PR if you wish though.

**Comment by trkoch** [comment]
> We should spend our time fixing those mysterious "issues". Not adding layers of special-case options.

I agree  – when meant as a general guideline. However, we should not forget that this is not Vim in its natural habitat and some concessions or compromises may be worth to consider when they improve overall usability from the users point of view.

> au WinEnter * startinsert

I see how this can replace `defaultMode`. From what I observe, `resetModeOnActiveEditorChange` controls behavior specific to this extension.

`true` (default):
* Editor A: Press `i`. Editor A is in insert mode.
* Switch to Editor B (e.g. via go to file): Editor B is in normal mode.
* Switch back to Editor B: Editor B is in normal mode.

`true` (default):
* Editor A: Press `i`. Editor A is in insert mode.
* Switch to Editor B (e.g. via go to file): Editor B is in normal mode.
* Switch back to Editor B: Editor B is in normal mode.

`false`:
* Editor A: Press `i`. Editor A is in insert mode.
* Switch to Editor B (e.g. via go to file): Editor B is in insert mode.
* Switch back to Editor B: Editor B is in insert mode.

Taking a step back here, I believe the underlying issue here is that in practice (I suspect) many users spend more time in insert mode to use features not available in normal/visual mode (mouse support, multi-cursor). Which is why setting the option above to `false` may be considered helpful by some.

**Comment by justinmk** [comment]
> #794, strikingly one of the most popular feature request.

+1 for #794 , it's more generalized.

> I see how this can replace `defaultMode`. From what I observe, `resetModeOnActiveEditorChange` controls behavior specific to this extension.

which parts of your table are not achieved by the `WinEnter` autocmd?

**Comment by theol0403** [comment]
Replaced by https://github.com/vscode-neovim/vscode-neovim/pull/1050 and potentially https://github.com/vscode-neovim/vscode-neovim/pull/1050. 

Use `au BufEnter * stopinsert` to leave insert mode when switching editors, and use `au BufEnter * startinsert` to enter insert mode when opening vscode or switching editors.
